File: Algorithms/15.3sum/15.3sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#
# https://leetcode.com/problems/3sum/description/
#
# algorithms
# Medium (23.14%)
# Total Accepted:    470.2K
# Total Submissions: 2M
# Testcase Example:  '[-1,0,1,2,-1,-4]'
#
# Given an array nums of n integers, are there elements a, b, c in nums such
# that a + b + c = 0? Find all unique triplets in the array which gives the sum
# of zero.
# 
# Note:
# 
# The solution set must not contain duplicate triplets.
# 
# Example:
# 
# 
# Given array nums = [-1, 0, 1, 2, -1, -4],
# 
# A solution set is:
# [
# ⁠ [-1, 0, 1],
# ⁠ [-1, -1, 2]
# ]
# 
# 
#
class Solution:
    def threeSum(self, nums: 'List[int]') -> 'List[List[int]]':
        res = []
        nums.sort()
        length = len(nums)
        for i in xrange(length-2):
            if nums[i]>0:break
            if i>0 and nums[i]==nums[i-1]:continue
            
            l,r = i+1,length-1
            while l<r:
                total = nums[i]+nums[l]+nums[r]

                if total<0:
                    l+=1
                elif total>0:
                    r-=1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l<r and nums[l]==nums[l+1]:
                        l+=1
                    while l<r and nums[r]==nums[r-1]:
                        r-=1
                    l+=1
                    r-=1
        return res
        

File: Algorithms/15.3sum/15.3sum.go
func threeSum(nums []int) [][]int {
    results := [][]int{}
	n := len(nums)
	if n == 0 || n < 3 {
		return results
	}
	sort.Ints(nums)  //首先，对数组进行排序
	for i := 0; i < n-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {  //如果相邻两个数相等
			continue
		}
		target := -nums[i]
		left := i + 1  //左指针
		right := n - 1  //右指针
		for left < right {
			sum := nums[left] + nums[right]
			if sum == target {
				results = append(results, []int{nums[left], nums[right], nums[i]})
				left++
				right--
				for left < right && nums[left] == nums[left-1] {
					left++
				}
				for left < right && nums[right] == nums[right+1] {
					right--
				}
			} else if sum > target {
				right--
			} else if sum < target {
				left++
			}
		}

	}
	return results
}
